pkg/apis/migration/migration.go

```diff
@@ -92,82 +92,87 @@ const (
 	CRDPlural = "cnsvspherevolumemigrations"
 )
 
-// onceForVolumeMigrationService is used for initializing the VolumeMigrationService singleton.
-var onceForVolumeMigrationService sync.Once
-
-// volumeMigrationInstance is instance of volumeMigration and implements interface for VolumeMigrationService
-var volumeMigrationInstance *volumeMigration
-
-var volumeMigrationServiceInitErr error
+var (
+	// volumeMigrationInstance is instance of volumeMigration and implements interface for VolumeMigrationService
+	volumeMigrationInstance *volumeMigration
+	// volumeMigrationInstanceLock is used for handling race conditions during read, write on volumeMigrationInstance
+	volumeMigrationInstanceLock = &sync.RWMutex{}
+)
 
 // GetVolumeMigrationService returns the singleton VolumeMigrationService
 func GetVolumeMigrationService(ctx context.Context, volumeManager *cnsvolume.Manager, cnsConfig *cnsconfig.Config) (VolumeMigrationService, error) {
 	log := logger.GetLogger(ctx)
-	onceForVolumeMigrationService.Do(func() {
-		log.Info("Initializing volume migration service...")
-		volumeMigrationInstance = &volumeMigration{
-			volumePathToVolumeID: sync.Map{},
-			volumeManager:        volumeManager,
-			cnsConfig:            cnsConfig,
-		}
-		volumeMigrationServiceInitErr = k8s.CreateCustomResourceDefinitionFromSpec(ctx, CRDName, CRDSingular, CRDPlural,
-			reflect.TypeOf(migrationv1alpha1.CnsVSphereVolumeMigration{}).Name(), migrationv1alpha1.SchemeGroupVersion.Group, migrationv1alpha1.SchemeGroupVersion.Version, apiextensionsv1beta1.ClusterScoped)
-		if volumeMigrationServiceInitErr != nil {
-			log.Errorf("failed to create volume migration CRD. Error: %v", volumeMigrationServiceInitErr)
-			return
-		}
-		config, volumeMigrationServiceInitErr := k8s.GetKubeConfig(ctx)
-		if volumeMigrationServiceInitErr != nil {
-			log.Errorf("failed to get kubeconfig. err: %v", volumeMigrationServiceInitErr)
-			return
-		}
-		volumeMigrationInstance.k8sClient, volumeMigrationServiceInitErr = k8s.NewClientForGroup(ctx, config, CRDGroupName)
-		if volumeMigrationServiceInitErr != nil {
-			log.Errorf("failed to create k8sClient. Err: %v", volumeMigrationServiceInitErr)
-			return
-		}
-		go func() {
-			log.Debugf("Starting Informer for cnsvspherevolumemigrations")
-			informer, err := k8s.GetDynamicInformer(ctx, migrationv1alpha1.SchemeGroupVersion.Group, migrationv1alpha1.SchemeGroupVersion.Version, "cnsvspherevolumemigrations", "")
-			if err != nil {
-				log.Errorf("failed to create dynamic informer for volume migration CRD. Err: %v", err)
-				return
+	volumeMigrationInstanceLock.RLock()
+	if volumeMigrationInstance == nil {
+		volumeMigrationInstanceLock.RUnlock()
+		volumeMigrationInstanceLock.RLock()
```

Comment:
This should be `volumeMigrationInstanceLock.Lock()`. Let me know if you dont understand why.

Reply:
Correct. Got it